### PR TITLE
Support for LTS15 of stackage and ghc88

### PR DESCRIPTION
**Pull request by junjihashimoto**
lts 15 of stackage is released.
This pr fixes https://github.com/hasktorch/hasktorch/issues/276.
I think stack's build works.
But nix's build requires gcc-9 which does not work with libtorch-1.4.

cabal works without nix.
https://github.com/hasktorch/hasktorch/pull/282/checks?check_run_id=450296535

### CODE REVIEW

**Comment by junjihashimoto** [comment]
https://hackage.haskell.org/package/oset
oset does not support base-4.13.

**Comment by junjihashimoto** [comment]
https://github.com/rcook/oset/pull/13

**Comment by tscholak** [comment]
@junjihashimoto I've been trying in parallel to fix a few additional things for this: https://github.com/tscholak/hasktorch/tree/feature/ghc88

**Comment by tscholak** [comment]
that includes oset and that nasty reduction stack overflow. please have a look or merge it into yours

**Comment by tscholak** [comment]
some jobs are still failing. I'm seeing things like this:
```
2020-02-28T02:41:22.1816980Z warning: unable to download 'https://hasktorch.cachix.org/xa9grk939csfmmd6bi2rxm4mg64iayjx.narinfo': HTTP error 502; retrying in 331 ms
2020-02-28T02:41:22.7511890Z warning: unable to download 'https://hasktorch.cachix.org/xa9grk939csfmmd6bi2rxm4mg64iayjx.narinfo': HTTP error 502; retrying in 629 ms
2020-02-28T02:41:23.4831440Z warning: unable to download 'https://hasktorch.cachix.org/xa9grk939csfmmd6bi2rxm4mg64iayjx.narinfo': HTTP error 502; retrying in 1102 ms
2020-02-28T02:41:24.7895650Z warning: unable to download 'https://hasktorch.cachix.org/xa9grk939csfmmd6bi2rxm4mg64iayjx.narinfo': HTTP error 502; retrying in 2312 ms
2020-02-28T02:41:27.2557780Z error: unable to download 'https://hasktorch.cachix.org/xa9grk939csfmmd6bi2rxm4mg64iayjx.narinfo': HTTP error 502
2020-02-28T02:41:27.2558720Z (use '--show-trace' to show detailed location information)
2020-02-28T02:41:27.2761750Z ##[error]Action failed with error: Error: The process 'nix-build' failed with exit code 1
2020-02-28T02:41:27.2780060Z (node:2687) UnhandledPromiseRejectionWarning: Error: The process 'nix-build' failed with exit code 1
2020-02-28T02:41:27.2780600Z     at ExecState._setResult (/Users/runner/runners/2.165.2/work/_actions/cachix/cachix-action/v2/node_modules/@actions/exec/lib/toolrunner.js:548:25)
2020-02-28T02:41:27.2781460Z     at ExecState.CheckComplete (/Users/runner/runners/2.165.2/work/_actions/cachix/cachix-action/v2/node_modules/@actions/exec/lib/toolrunner.js:531:18)
2020-02-28T02:41:27.2782670Z     at ChildProcess.<anonymous> (/Users/runner/runners/2.165.2/work/_actions/cachix/cachix-action/v2/node_modules/@actions/exec/lib/toolrunner.js:431:27)
2020-02-28T02:41:27.2783000Z     at ChildProcess.emit (events.js:210:5)
2020-02-28T02:41:27.2783130Z     at maybeClose (internal/child_process.js:1021:16)
2020-02-28T02:41:27.2783230Z     at Socket.<anonymous> (internal/child_process.js:430:11)
2020-02-28T02:41:27.2783320Z     at Socket.emit (events.js:210:5)
2020-02-28T02:41:27.2783580Z     at Pipe.<anonymous> (net.js:659:12)
2020-02-28T02:41:27.2783700Z (node:2687) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
2020-02-28T02:41:27.2784920Z (node:2687) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
Other jobs are just being killed after a while

**Comment by junjihashimoto** [comment]
@tscholak 
I can not fix nasty reduction stack overflow, too.
And I commented out it.
https://github.com/hasktorch/hasktorch/pull/282/files#diff-5ab82fdbc38d8ce2310fe77f404194ecR1841-R1848

**Comment by tscholak** [comment]
I think I have fixed it, though: https://github.com/tscholak/hasktorch/commit/78923ff4631998afd64c9264c93eb829da5cfcc2

**Comment by tscholak** [comment]
here's the oset patch: https://github.com/tscholak/hasktorch/commit/f3a7d1b8ba058bcf7aaf4da3ca677fbdff46740e

**Comment by junjihashimoto** [comment]
Thank you for your help.

**Comment by junjihashimoto** [comment]
I revert {-# OPTIONS_GHC -freduction-depth=0 #-} to avoid infinit loop of ghc.
CI takes 3 hours with the option, and  it does not finish.

**Comment by junjihashimoto** [comment]
ghc883 does not fix following inference.
https://github.com/hasktorch/hasktorch/pull/282/checks?check_run_id=488090361#step:5:5829
```
src/Torch/Typed/NN.hs:314:41: error:
    • Could not deduce: (1 <=? stride) ~ 'True
        arising from a use of ‘Torch.Typed.NN.conv1d’
      from the context: (All
                           KnownNat
                           '[stride, padding, inputChannelSize, outputChannelSize, kernelSize,
                             inputSize, batchSize, outputSize],
                         ConvSideCheck inputSize kernelSize stride padding outputSize)
        bound by the instance declaration
        at src/Torch/Typed/NN.hs:(304,3)-(313,244)
    • In the expression:
        Torch.Typed.NN.conv1d @stride @padding model input
      In an equation for ‘forward’:
          forward model (input, Proxy, Proxy)
            = Torch.Typed.NN.conv1d @stride @padding model input
      In the instance declaration for
        ‘HasForward
           (Conv1d inputChannelSize outputChannelSize kernelSize dtype device)
           (Tensor device dtype '[batchSize, inputChannelSize, inputSize],
```

**Comment by junjihashimoto** [comment]
This means plugins of ghc-typelits-* do not work?

**Comment by tscholak** [comment]
hm, I can't believe that. Does this maybe have to do with the constraints you commented because of that infinitely recursive check problem?
I cannot investigate this further right now myself, maybe on the weekend

**Comment by junjihashimoto** [comment]
cabal's packages of typelits is different from nix's one.
https://gist.github.com/junjihashimoto/b70b42302507c95c7bdc5c2d6b0b14a8

**Comment by junjihashimoto** [comment]
Bingo!

**Comment by junjihashimoto** [comment]
@tscholak 
I've fix pr for stack and cabal by pinned plugins.

**Comment by tscholak** [comment]
that's awesome @junjihashimoto.
there are some new and weird build errors in the nix-linux build:
```
  Building test suite 'test-rsa' for RSA-2.3.1..
35612
  [1 of 2] Compiling Codec.Crypto.RSA.Pure ( src/Codec/Crypto/RSA/Pure.hs, dist/build/test-rsa/test-rsa-tmp/Codec/Crypto/RSA/Pure.o )
35613
  [47 of 48] Compiling All              ( test/All.hs, dist/build/spec/spec-tmp/All.o )
35614
  Linking dist/build/spec/spec ...
35615
  
35616
  src/Codec/Crypto/RSA/Pure.hs:115:25: error:
35617
      • Could not deduce (MonadFail m) arising from a use of ‘fail’
35618
        from the context: (Monad m, Show a)
35619
          bound by the type signature for:
35620
                     failOnError :: forall (m :: * -> *) a b.
35621
                                    (Monad m, Show a) =>
35622
                                    Either a b -> m b
35623
          at src/Codec/Crypto/RSA/Pure.hs:114:1-53
35624
        Possible fix:
35625
          add (MonadFail m) to the context of
35626
            the type signature for:
35627
              failOnError :: forall (m :: * -> *) a b.
35628
                             (Monad m, Show a) =>
35629
                             Either a b -> m b
35630
      • In the expression: fail (show e)
35631
        In an equation for ‘failOnError’:
35632
            failOnError (Left e) = fail (show e)
35633
      |
35634
  115 | failOnError (Left e)  = fail (show e)
35635
      |                         ^^^^^^^^^^^^^
```

**Comment by tscholak** [comment]
```

**Comment by tscholak** [comment]
dist/build/libHSlibtorch-ffi-1.4.0.0-6iVIwOZm2nvBEbSLmylc95-ghc8.8.2.a: hPutBuf: resource exhausted (No space left on device)
54701
  error: writing to file: No space left on device
54702
  ##[error]Action failed with error: Error: The process 'nix-build' failed with exit code 1
```
this is so frustrating :/

**Comment by junjihashimoto** [comment]
@tscholak 
Yeah. So I've splited the nix-build into 3parts.

**Comment by junjihashimoto** [comment]
Oh, we can remove std=c++11 options.
https://gcc.gnu.org/projects/cxx-status.html
image

**Comment by junjihashimoto** [comment]
nix-macos:

image
https://github.com/hasktorch/hasktorch/runs/535222093

**Comment by junjihashimoto** [comment]
stack-macos:
relocation error
image

**Comment by junjihashimoto** [comment]
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/haskell-modules/generic-builder.nix#L388

**Comment by junjihashimoto** [comment]
https://github.com/NixOS/nixpkgs/pull/83605

File: hasktorch/src/Torch/Typed/Functional.hs [hunk 1]
```diff
@@ -1852,11 +1852,12 @@ constantPadNd1d value input = unsafePerformIO $ ATen.cast3
 type ConvSideCheck (inputSize :: Nat) (kernelSize :: Nat) (stride :: Nat) (padding :: Nat) (outputSize :: Nat) =
   (
     -- kernel size and stride must be > 0
-    kernelSize >= 1, stride >= 1
+    1 <= kernelSize, 1 <= stride
     -- kernel size can't be greater than actual input size
-  , ((inputSize + (2 * padding)) + 1) >= kernelSize
+    -- ToDo: Do not use '>=' on constraint to avoid reduction-stack-overflow.
+  , (kernelSize - 1) <= (inputSize + (2 * padding))
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Typed/Functional.hs]: This `>=` seems like an obstacle.
When we use `>=` , deduction-error occurs.

Comment (junjihashimoto) [reply hasktorch/src/Torch/Typed/Functional.hs]: In addition to this `>=`, `  , `kernelSize <= (inputSize + (2 * padding) + 1)` becomes deduction-error too, so I use `(kernelSize -1) <= (inputSize + (2 * padding))` .

Comment (junjihashimoto) [reply hasktorch/src/Torch/Typed/Functional.hs]: Not ghc-typelits-natnormalise-0.7.2 but 0.7 allows the deduction-error.(I do not know why.)

**Comment by junjihashimoto** [comment]
There were some problems.

- libtorch-1.4 has compilation-errors with gcc-9.
  - https://github.com/stites/pytorch-world/pull/18
- We need to bump version of nixpkgs to use ghc883, but the nixpkgs changes the way to create symlink of shared libraries, the change causes the error of confliction of libiomp5.dylib.
  - https://github.com/stites/pytorch-world/pull/19
  - https://github.com/stites/pytorch-world/pull/20
- When we use '>=', ghc-typelits-natnormalise causes the error of recursive reduction.
- doctest of hasktorch causes relocaiton error on macos. For now, I disable the test.

**Comment by austinvhuang** [comment]
Yikes thanks for all this! FYI libtorch 1.5 is landing very soon, maybe the gcc errors will be resolved 🤞

File: libtorch-ffi/libtorch-ffi.cabal [hunk 2]
```diff
@@ -116,20 +116,11 @@ test-suite spec
...
                      , hspec
                      , hspec-discover
                      , safe-exceptions
   if os(darwin)
     ld-options: -Wl,-keep_dwarf_unwind
-    ghc-options:       -optc-std=c++11 -optc-xc++
-  else
-    ghc-options:       -optc-std=c++11
-    cc-options:        -std=c++11
-  cxx-options:       -std=c++11
```
Comment (tscholak) [inline libtorch-ffi/libtorch-ffi.cabal:133]: yay! I always hated these

**Comment by junjihashimoto** [comment]
@tscholak @austinvhuang 
Thank you for your reviewing.
I am looking forward to libtorch-1.5. 
I wanted to make a simple example for the gcc-error and feed it back to GCC, but I've not been able to reproduce the error.
I was troubled by macos's behavior, but I'm glad I could solve it.

**Comment by jrp2014** [comment]
Thanks for this!

**Pull request merged**